overlord/devicestate/devicemgr.go

```diff
@@ -0,0 +1,373 @@
+		return nil, fmt.Errorf("cannot read device key pair: %v", err)
+	}
+	return privKey, nil
+}
+
+type serialSetup struct {
+	SerialRequest string `json:"serial-request"`
+}
+
+type requestIDResp struct {
+	RequestID string `json:"request-id"`
+}
+
+func prepareSerialRequest(t *state.Task, privKey asserts.PrivateKey, device *auth.DeviceState, client *http.Client) (string, error) {
+	st := t.State()
+	st.Unlock()
+	defer st.Lock()
+	resp, err := client.Get(serialRequestURL)
+	if err != nil {
+		st.Lock()
+		t.Errorf("cannot retrieve request-id for making a request for a serial: %v", err)
+		st.Unlock()
+		return "", &state.Retry{After: 60 * time.Second}
+	}
+	if resp.StatusCode != 200 {
+		st.Lock()
+		t.Errorf("cannot retrieve request-id for making a request for a serial: unexpected status %d", resp.StatusCode)
+		st.Unlock()
+		return "", &state.Retry{After: 60 * time.Second}
+	}
+
+	dec := json.NewDecoder(resp.Body)
+	var requestID requestIDResp
+	err = dec.Decode(&requestID)
+	if err != nil { // assume broken i/o
+		st.Lock()
+		t.Errorf("cannot read response with request-id for making a request for a serial: %v", err)
+		st.Unlock()
+		return "", &state.Retry{After: 60 * time.Second}
+	}
+
+	encodedPubKey, err := asserts.EncodePublicKey(privKey.PublicKey())
+	if err != nil {
+		return "", fmt.Errorf("internal error: cannot encode device public key: %v", err)
+
+	}
+
+	serialReq, err := asserts.SignWithoutAuthority(asserts.SerialRequestType, map[string]interface{}{
+		"brand-id":   device.Brand,
+		"model":      device.Model,
+		"request-id": requestID.RequestID,
+		"device-key": string(encodedPubKey),
+	}, nil, privKey) // XXX: fill body with some agreed hardware details
+	if err != nil {
+		return "", err
+	}
+
+	return string(asserts.Encode(serialReq)), nil
+}
+
+var errPoll = errors.New("serial-request accepted, poll later")
+
+func submitSerialRequest(t *state.Task, serialRequest string, client *http.Client) (*asserts.Serial, error) {
+	st := t.State()
+	st.Unlock()
+	defer st.Lock()
+	resp, err := client.Post(serialRequestURL, asserts.MediaType, bytes.NewBufferString(serialRequest))
+	if err != nil {
+		st.Lock()
+		t.Errorf("cannot POST request for a serial: %v", err)
+		st.Unlock()
+		return nil, &state.Retry{After: 60 * time.Second}
+	}
+
+	switch resp.StatusCode {
+	case 200, 201:
+	case 202:
+		return nil, errPoll
+	default:
+		st.Lock()
+		t.Errorf("cannot POST request for a serial: unexpected status %d", resp.StatusCode)
+		st.Unlock()
+		return nil, &state.Retry{After: 60 * time.Second}
+	}
+
+	// decode body with serial assertion
+	dec := asserts.NewDecoder(resp.Body)
+	got, err := dec.Decode()
+	if err != nil { // assume broken i/o
+		st.Lock()
+		t.Errorf("cannot read response to request for a serial: %v", err)
+		st.Unlock()
+		return nil, &state.Retry{After: 60 * time.Second}
+	}
+
+	serial, ok := got.(*asserts.Serial)
+	if !ok {
+		return nil, fmt.Errorf("request for a serial response did not return a serial assertion")
```

Comment:
Actually, we have an assertion type at hand in string form, right? That'd be better than the Go type.